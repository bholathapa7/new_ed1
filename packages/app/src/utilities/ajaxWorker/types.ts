import { AjaxRequest } from 'rxjs/ajax';

export const enum AJAX_METHOD {
  GET = 'GET',
  POST = 'POST',
  PATCH = 'PATCH',
  DELETE = 'DELETE',
  BASIC = 'BASIC',
  RAW = 'RAW',
}

export type AjaxMethod =
  AJAX_METHOD.GET | AJAX_METHOD.POST | AJAX_METHOD.PATCH | AJAX_METHOD.DELETE | AJAX_METHOD.BASIC | AJAX_METHOD.RAW;

export interface AjaxParams {
  url: string;
  headers?: object;
  body?: any;
}

export interface MsgFromMainThread {
  ajaxParams: AjaxParams | (string | AjaxRequest);
  method: AjaxMethod;
}

export interface BasicAjaxParams {
  urlOrRequest?: string | AjaxRequest;
}

export interface AjaxWorkerEvent extends MessageEvent {
  data: AjaxParams;
}
