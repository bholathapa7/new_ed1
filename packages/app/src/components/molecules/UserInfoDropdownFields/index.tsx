import React, { FC, ReactNode } from 'react';
import styled, { CSSObject } from 'styled-components';

import * as T from '^/types';

import countries from '^/constants/countries';
import palette from '^/constants/palette';

import { l10n } from '^/utilities/l10n';

import RawDropdown, {
  Option as DropdownOption,
  Props as DropdownProps,
} from '^/components/atoms/Dropdown/1';
import withL10n, { L10nProps } from '^/components/atoms/WithL10n';

import Text from './text';

interface CustomStyleProps {
  readonly customStyle?: CSSObject;
}
const Root = styled.li<CustomStyleProps>({
  display: 'block',

  ':not(:first-of-type)': {
    marginTop: '20px',
  },
}, ({ customStyle }) => (
  customStyle ? { ...customStyle } : {}
));

const Wrapper = styled.div({
  position: 'relative',
});

const Label = styled.label<CustomStyleProps>({
  display: 'block',
  marginBottom: '10px',

  color: palette.darkBlack.toString(),
}, ({ customStyle }) => (
  customStyle ? { ...customStyle } : {}
));

const Message = styled.p({
  display: 'block',
  marginTop: '10px',

  lineHeight: 1.3,
  fontSize: '14px',
});

const ErrorMessage = styled(Message)({
  color: palette.error.toString(),
});

const Dropdown = styled(RawDropdown)<DropdownProps>({
  width: '100%',
});
Dropdown.displayName = 'Dropdown';

export type UserInfoDropdownFieldsKind =
  | 'country'
  | 'purpose'
  | 'language'
  ;

const languageTypeToLanguageText: Record<T.Language, string> = {
  [T.Language.EN_US]: 'English',
  [T.Language.KO_KR]: '한국어',
};

const kindToLabel: Record<UserInfoDropdownFieldsKind, (lang: T.Language) => string> = {
  country: (lang) => l10n(Text.countryLabel, lang),
  purpose: (lang) => l10n(Text.purposeLabel, lang),
  language: (lang) => l10n(Text.languageLabel, lang),
};

type Options = Array<DropdownOption>;
const kindToOptions: Record<UserInfoDropdownFieldsKind, (lang: T.Language) => Options> = ({
  country: (lang) => countries[lang]
    .map((text, index) => ({
      text,
      value: countries[T.Language.EN_US][index],
    })),
  purpose: (lang) => Object.keys(Text.purposeOptions)
    .filter((key) => key !== 'default')
    .map((key: keyof (typeof Text.purposeOptions)) => ({
      text: Text.purposeOptions[key][lang],
      value: key,
    })),
  language: () => Object.keys(languageTypeToLanguageText)
    .map((key) => ({
      text: languageTypeToLanguageText[key as T.Language],
      value: key,
    })),
});

const kindToPlaceHolder: Record<UserInfoDropdownFieldsKind, (lang: T.Language) => string> = ({
  country: (lang) => l10n(Text.countryOptions.default, lang),
  purpose: (lang) => l10n(Text.purposeOptions.default, lang),
  language: (lang) => l10n(Text.languageOptions.default, lang),
});

const kindToHeight: Record<UserInfoDropdownFieldsKind, (lang: T.Language) => string> = ({
  country: (): string => '178px',
  purpose: (): string => 'auto',
  language: (): string => 'auto',
});

export interface Props {
  readonly kind: UserInfoDropdownFieldsKind;
  readonly value?: string;
  readonly error?: string;
  readonly rootStyle?: CSSObject;
  readonly labelStyle?: CSSObject;
  onChange(kind: UserInfoDropdownFieldsKind, value: string): void;
}

/**
 * @author Junyoung Clare Jang
 * @desc Wed Oct 24 18:28:32 2018 UTC
 * @desc field for sign up
 */
const UserInfoDropdownFields: FC<Props & L10nProps> = ({
  error, kind, value, language, rootStyle, onChange,
}) => {
  const handleDropdownChange: ({ value: dropdownValue }: DropdownOption) => void = ({ value: dropdownValue }) => {
    onChange(kind, dropdownValue as string);
  };

  const withError: boolean = error !== undefined;
  const errorMessage: ReactNode =
    withError ?
      <ErrorMessage>{`* ${error}`}</ErrorMessage> :
      undefined;

  const labelText: string = kindToLabel[kind](language);

  const options: Options = kindToOptions[kind](language);
  const placeHolder: string = kindToPlaceHolder[kind](language);
  const height: string = kindToHeight[kind](language);

  return (
    <Root customStyle={rootStyle}>
      <Wrapper>
        <Label>{labelText}</Label>
        <Dropdown
          value={value}
          placeHolder={placeHolder}
          options={options}
          height={height}
          zIndex={1}
          onClick={handleDropdownChange}
          error={withError}
          isSearchEnable={kind === 'country'}
        />
        {errorMessage}
      </Wrapper>
    </Root>
  );
};
export default withL10n(UserInfoDropdownFields);
