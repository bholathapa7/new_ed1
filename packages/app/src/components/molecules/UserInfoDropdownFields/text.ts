import { Language } from '^/types';

export default {
  countryLabel: {
    [Language.KO_KR]: '국가',
    [Language.EN_US]: 'Country',
  },
  purposeLabel: {
    [Language.KO_KR]: '용도',
    [Language.EN_US]: 'Purpose',
  },
  languageLabel: {
    [Language.KO_KR]: '언어',
    [Language.EN_US]: 'Language',
  },

  countryPlaceholder: {
    [Language.KO_KR]: '국가를 선택해주세요',
    [Language.EN_US]: 'Select your country',
  },
  purposePlaceholder: {
    [Language.KO_KR]: '사용 용도를 선택해주세요',
    [Language.EN_US]: 'Select your purpose',
  },
  languagePlaceholder: {
    [Language.KO_KR]: '사용하시는 언어를 선택해주세요',
    [Language.EN_US]: 'Select your language',
  },

  countryOptions: {
    default: {
      [Language.KO_KR]: '국가를 선택해주세요',
      [Language.EN_US]: 'Select your country',
    },
  },

  purposeOptions: {
    construction: {
      [Language.KO_KR]: '건설',
      [Language.EN_US]: 'Construction',
    },
    survey: {
      [Language.KO_KR]: '측량',
      [Language.EN_US]: 'Survey',
    },
    urbanPlanning: {
      [Language.KO_KR]: '도시계획',
      [Language.EN_US]: 'Urban Planning',
    },
    disasterResponse: {
      [Language.KO_KR]: '재난재해 대비/대책',
      [Language.EN_US]: 'Disaster Response',
    },
    inspection: {
      [Language.KO_KR]: '시설물 점검',
      [Language.EN_US]: 'Inspection',
    },
    droneService: {
      [Language.KO_KR]: '드론 서비스업',
      [Language.EN_US]: 'Drone Service',
    },
    others: {
      [Language.KO_KR]: '기타',
      [Language.EN_US]: 'Others',
    },
    default: {
      [Language.KO_KR]: '용도를 선택해주세요',
      [Language.EN_US]: 'Select your purpose',
    },
  },

  languageOptions: {
    default: {
      [Language.KO_KR]: '언어를 선택해주세요',
      [Language.EN_US]: 'Select your language',
    },
  },
};
