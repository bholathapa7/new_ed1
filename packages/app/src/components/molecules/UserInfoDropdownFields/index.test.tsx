import * as ReactTesting from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import * as T from '^/types';

import {
  DDMMockStore,
  commonAfterEach,
  createDDMMockStore,
} from '^/utilities/test-util';

import UserInfoDropdownFields, { Props } from './';

describe('UserInfoDropdownFields', () => {
  const createProps: () => Props = () => ({
    kind: 'country',
    value: 'United States',
    onChange: jest.fn(),
  });
  let props: Props;
  let store: DDMMockStore;
  let renderResult: ReactTesting.RenderResult;

  beforeEach(() => {
    store = createDDMMockStore(T.Language.KO_KR);
    props = createProps();
    renderResult = ReactTesting.render(
      <Provider store={store}>
        <UserInfoDropdownFields {...props} />
      </Provider>,
    );
  });

  afterEach(ReactTesting.cleanup);
  afterEach(commonAfterEach);

  it('should have appropriate label for each kinds', () => {
    const label: HTMLElement =
      renderResult.container.getElementsByTagName('label')[0];

    renderResult.rerender(
      <Provider store={store}>
        <UserInfoDropdownFields {...props} kind='country' />
      </Provider>,
    );
    expect(label.textContent).toMatchSnapshot('country label');

    renderResult.rerender(
      <Provider store={store}>
        <UserInfoDropdownFields {...props} kind='purpose' />
      </Provider>,
    );
    expect(label.textContent).toMatchSnapshot('purpose label');

    renderResult.rerender(
      <Provider store={store}>
        <UserInfoDropdownFields {...props} kind='language' />
      </Provider>,
    );
    expect(label.textContent).toMatchSnapshot('language label');
  });

  it('should have appropriate placeholder for each kinds', () => {
    const dropdownButton: HTMLElement =
      renderResult.getByTestId('dropdown-mainbutton');

    renderResult.rerender(
      <Provider store={store}>
        <UserInfoDropdownFields {...props} kind='country' value={undefined} />
      </Provider>,
    );
    expect(dropdownButton.textContent).toMatchSnapshot('country placeholder');

    renderResult.rerender(
      <Provider store={store}>
        <UserInfoDropdownFields {...props} kind='purpose' value={undefined} />
      </Provider>,
    );
    expect(dropdownButton.textContent).toMatchSnapshot('purpose placeholder');

    renderResult.rerender(
      <Provider store={store}>
        <UserInfoDropdownFields {...props} kind='language' value={undefined} />
      </Provider>,
    );
    expect(dropdownButton.textContent).toMatchSnapshot('language placeholder');
  });
});
