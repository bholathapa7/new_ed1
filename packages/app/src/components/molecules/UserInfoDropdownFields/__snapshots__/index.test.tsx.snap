// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfoDropdownFields should have appropriate label for each kinds: country label 1`] = `"국가"`;

exports[`UserInfoDropdownFields should have appropriate label for each kinds: language label 1`] = `"언어"`;

exports[`UserInfoDropdownFields should have appropriate label for each kinds: purpose label 1`] = `"용도"`;

exports[`UserInfoDropdownFields should have appropriate placeholder for each kinds: country placeholder 1`] = `"국가를 선택해주세요"`;

exports[`UserInfoDropdownFields should have appropriate placeholder for each kinds: language placeholder 1`] = `"언어를 선택해주세요"`;

exports[`UserInfoDropdownFields should have appropriate placeholder for each kinds: purpose placeholder 1`] = `"용도를 선택해주세요"`;
